File: 백준/Gold/12441. 약속장소 정하기 （Small）/약속장소 정하기 （Small）.py
# 약속장소 정하기(small)
import sys
input = sys.stdin.readline
import heapq as hq

def dijkstra(n):
    que = []
    local_visited = [0] * (N+1)
    dist = [float('inf')] * (N+1)
    dist[start[n]] = 0    
    hq.heappush(que, (0, start[n]))
    cur_speed = speed[n]
    
    while que:
        cost, cur = hq.heappop(que)

        if local_visited[cur]:
            continue

        local_visited[cur] = 1
        visited[cur] |= (1 << n)
        total[cur] = max(total[cur], cost)

        # 최소거리가 갱신 가능한 경우만 que에 넣는다
        for ncity, ncost in graph[cur]:
            if dist[ncity] > cost + ncost*cur_speed:
                dist[ncity] = cost + ncost*cur_speed
                hq.heappush(que, (dist[ncity], ncity))

T = int(input())

for i in range(T):
    # 도시, 친구, 도로의 숫자
    N, P, M = map(int,input().split()) 
    start = [0] * (N+1) # 친구 시작 도시 
    speed = [0] * (N+1) # 친구 이동속도
    graph = [[] for _ in range(N+1)] # 도로 연결그래프
    visited = [1] * (N+1) # 비트마스킹으로 방문 확인
    total = [0] * (N+1) # 도시별 최대 도달시간 저장
    
    for j in range(P):
        X, V = map(int,input().split())
        start[j+1] = X
        speed[j+1] = V

    for j in range(M):
        roads = list(map(int,input().split()))

        dist = roads[0]
        links = roads[1]

        # 주어진 도로들의 연결정보 저장
        for k in range(links):
            n1 = roads[2+k]
            for m in range(k+1, links):
                n2 = roads[2+m]
                graph[n1].append((n2, dist*(m-k)))
                graph[n2].append((n1, dist*(m-k)))

    for j in range(1, P+1):
        dijkstra(j)
            
    minn = float('inf')
    for j in range(1, N+1):
        if visited[j] != (1 << (P+1)) - 1:
            continue

        if minn > total[j]:
            minn = total[j]

    if minn == float('inf'):
        print(f'Case #{i+1}: -1')
    else:
        print(f'Case #{i+1}: {minn}')

File: 백준/Gold/12441. 약속장소 정하기 （Small）/약속장소 정하기 （Small）.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node implements Comparable<Node> {
        int city, time;
        Node(int city, int time) {
            this.city = city;
            this.time = time;
        }

        public int compareTo(Node o) {
            return Integer.compare(this.time, o.time);
        }
    }

    static int N, P, M;
    static List<List<int[]>> graph;
    static int[] start;
    static int[] speed;
    static int[] total;
    static int[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        for (int tc = 1; tc <= T; tc++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            P = Integer.parseInt(st.nextToken());
            M = Integer.parseInt(st.nextToken());

            start = new int[P + 1];
            speed = new int[P + 1];
            total = new int[N + 1];
            visited = new int[N + 1];
            Arrays.fill(visited, 1);  // 초기 비트마스크 값

            graph = new ArrayList<>();
            for (int i = 0; i <= N; i++) {
                graph.add(new ArrayList<>());
            }

            // 친구들의 시작 위치와 속도
            for (int i = 1; i <= P; i++) {
                st = new StringTokenizer(br.readLine());
                int X = Integer.parseInt(st.nextToken());
                int V = Integer.parseInt(st.nextToken());
                start[i] = X;
                speed[i] = V;
            }

            // 도로 정보
            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                int dist = Integer.parseInt(st.nextToken());
                int links = Integer.parseInt(st.nextToken());
                int[] nodes = new int[links];
                for (int j = 0; j < links; j++) {
                    nodes[j] = Integer.parseInt(st.nextToken());
                }
                for (int j = 0; j < links; j++) {
                    for (int k = j + 1; k < links; k++) {
                        int cost = dist * (k - j);
                        graph.get(nodes[j]).add(new int[]{nodes[k], cost});
                        graph.get(nodes[k]).add(new int[]{nodes[j], cost});
                    }
                }
            }

            for (int i = 1; i <= P; i++) {
                dijkstra(i);
            }

            int answer = Integer.MAX_VALUE;
            for (int i = 1; i <= N; i++) {
                if (visited[i] != (1 << (P + 1)) - 1) continue;
                answer = Math.min(answer, total[i]);
            }

            if (answer == Integer.MAX_VALUE) {
                System.out.println("Case #" + tc + ": -1");
            } else {
                System.out.println("Case #" + tc + ": " + answer);
            }
        }
    }

    static void dijkstra(int friendIdx) {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        boolean[] localVisited = new boolean[N + 1];
        int[] dist = new int[N + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);

        int startCity = start[friendIdx];
        int curSpeed = speed[friendIdx];
        pq.offer(new Node(startCity, 0));
        dist[startCity] = 0;

        while (!pq.isEmpty()) {
            Node cur = pq.poll();
            if (localVisited[cur.city]) continue;
            localVisited[cur.city] = true;

            if ((visited[cur.city] & (1 << friendIdx)) == 0) {
                visited[cur.city] |= (1 << friendIdx);
                total[cur.city] = Math.max(total[cur.city], cur.time);
            }

            for (int[] next : graph.get(cur.city)) {
                int nextCity = next[0];
                int moveCost = next[1] * curSpeed;
                if (dist[nextCity] > cur.time + moveCost) {
                    dist[nextCity] = cur.time + moveCost;
                    pq.offer(new Node(nextCity, dist[nextCity]));
                }
            }
        }
    }
}
